docs/standard/native-interop/tutorial-custom-marshaller.md

```diff
@@ -0,0 +1,278 @@
+            public byte IsFatal;    // The C++ bool is defined as a single byte.
+            public uint* Message;   // This could be as simple as a void*, but uint* is closer.
+        }
+
+        public static ErrorDataUnmanaged ConvertToUnmanaged(ErrorData managed)
+            => throw new NotImplementedException();
+
+        public static void Free(ErrorDataUnmanaged unmanaged)
+            => throw new NotImplementedException();
+    }
+}
+```
+
+The conversion from an `ErrorData` to an `ErrorDataUnmanaged`, which will mimic the shape of the unmanaged type, is now trivial with `Utf32StringMarshaller`. This is a "stateless" marshaller because the marshaller itself only consists of `static` functions.
+
+Marshalling of an `int` is unnecessary since its representation is identical in unmanaged and managed code. A `bool` value's binary representation isn't defined in .NET, so you use its current value to define a zero and non-zero value in the unmanaged type. Then, reuse your UTF-32 marshaller to convert the `string` field into a `uint*`.
+
+```csharp
+public static ErrorDataUnmanaged ConvertToUnmanaged(ErrorData managed)
+{
+    return new ErrorDataUnmanaged
+    {
+        Code = managed.Code,
+        IsFatal = (byte)(managed.IsFatalError ? 1 : 0),
+        Message = Utf32StringMarshaller.ConvertToUnmanaged(managed.Message),
+    };
+}
+```
+
+Recall that you're defining this marshaller as an "in", so you must clean up any allocations performed during the marshalling. The `int` and `bool` fields didn't allocate any memory, but the `Message` field did. Reuse `Utf32StringMarshaller` again to clean up the marshalled string.
+
+```csharp
+public static void Free(ErrorDataUnmanaged unmanaged)
+    => Utf32StringMarshaller.Free(unmanaged.Message);
+```
+
+Next, consider the case where multiple instances of `error_data` are returned. Typically, you use a collection to return multiple elements. Since this marshaller will be returning multiple elements, this is called the `MarshalMode.ElementOut` mode.
+
+```csharp
+namespace CustomMarshalling
+{
+    [CustomMarshaller(typeof(ErrorData), MarshalMode.ManagedToUnmanagedIn, typeof(ErrorDataMarshaller))]
+    [CustomMarshaller(typeof(ErrorData), MarshalMode.ElementOut, typeof(Out))]
+    internal static unsafe class ErrorDataMarshaller
+    {
+        //
+        // Other marshallers removed
+        //
+
+        public static class Out
+        {
+            public static ErrorData ConvertToManaged(ErrorDataUnmanaged unmanaged)
+                => throw new NotImplementedException();
+
+            public static void Free(ErrorDataUnmanaged unmanaged)
+                => throw new NotImplementedException();
+        }
+    }
+}
+```
+
+The conversion from `ErrorDataUnmanaged` to `ErrorData` is the inverse of what you did for the "in" mode in the preceding code. Remember that you also need to "clean up" any allocations that the unmanaged environment expected you to perform.
+
```

Comment:
This tutorial currently only shows p/invokes where the 'in' marshaller would be used. It might be helpful to also have a snippet with the p/invoke that would use this marshaller (and the one below). I think it'd also make the break/switch from talking about `ElementOut` to `ManagedToUnmanagedOut` a bit clearer.

Reply:
I thnk a snippet for non-element would be helpful too. Maybe something like:


>Now, consider the case where only one `error_data` instance is returned.
>
>[c++ and p/invoke snippet]
>
>For the marshaller used in this scenario, corresponding to the `MarshalMode.ManagedToUnmanagedOut` mode, you're going to do something special...